core/metrics-core-service/src/main/java/org/hawkular/metrics/core/service/MetricsServiceImpl.java

```diff
@@ -497,29 +498,36 @@ public void setDefaultTTL(int defaultTTL) {
     }
 
     @Override
-    public Observable<Metric<?>> findAllMetrics() {
-        return dataAccess.findAllMetricsInData()
-                .compose(new MetricFromFullDataRowTransformer(defaultTTL))
+    public Observable<MetricId<?>> findAllMetricIdentifiers() {
+        return dataAccess.findAllMetricIdentifiersInData()
+                .compose(new MetricIdentifierFromFullDataRowTransformer(defaultTTL))
                 .distinct();
     }
 
+    public <T> Observable.Transformer<MetricId<T>, Metric<T>> enrichToMetric() {
+        return t -> t
+                .flatMap(id -> dataAccess.findMetricInMetricsIndex(id)
+                        .compose(new MetricsIndexRowTransformer<>(id.getTenantId(), id.getType(), defaultTTL))
+                        .switchIfEmpty(dataAccess.findMetricInData(id) // This only verifies it exists..
+                                .compose(new MetricFromDataRowTransformer<>(id.getTenantId(), id.getType(), defaultTTL))));
+    }
+
     @Override
     public <T> Observable<Metric<T>> findMetric(final MetricId<T> id) {
-        return dataAccess.findMetricInMetricsIndex(id)
-                .compose(new MetricsIndexRowTransformer<>(id.getTenantId(), id.getType(), defaultTTL))
-                .switchIfEmpty(dataAccess.findMetricInData(id)
-                        .compose(new MetricFromDataRowTransformer<>(id.getTenantId(), id.getType(), defaultTTL)));
+        return Observable.just(id)
+                .compose(enrichToMetric());
     }
 
     @SuppressWarnings("unchecked")
     @Override
     public <T> Observable<Metric<T>> findMetrics(String tenantId, MetricType<T> metricType) {
         Observable<Metric<T>> setFromMetricsIndex = null;
-        Observable<Metric<T>> setFromData = dataAccess.findAllMetricsInData()
+        Observable<Metric<T>> setFromData = dataAccess.findAllMetricIdentifiersInData()
                 .doOnError(Throwable::printStackTrace)
                 .filter(row -> tenantId.equals(row.getString(0)))
                 .distinct()
-                .compose(new MetricFromFullDataRowTransformer(defaultTTL))
+                .compose(new MetricIdentifierFromFullDataRowTransformer(defaultTTL))
+                .flatMap(this::findMetric)
```

Comment:
I am not sure I understand what you are saying. I think the ```distinct``` call at line 528 is not needed. See [HWKMETRICS-715](https://issues.jboss.org/browse/HWKMETRICS-715). I think it should come after line 529.

We do not store metric tags in the data tables. If the metric id is found in the data tables, it is not necessarily in metrics_idx. Line 530 is doing the same thing as the call to ```findMetricsInMetricsIndex``` at line 539, except it is getting executed for every metric id.